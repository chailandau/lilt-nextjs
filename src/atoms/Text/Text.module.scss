.text {
    @include font(18, 24, 0.35);
    @include tablet {
        @include font(20, 28, 0.4);
    }
}
