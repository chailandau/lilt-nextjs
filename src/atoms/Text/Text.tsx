import { FC, ReactNode } from 'react';

import styles from './Text.module.scss';

export const textTags = ['p', 'span'] as const;

interface TextProps {
    /** Semantic tag text is wrapped in */
    as?: (typeof textTags)[number];
    /** Text content */
    children: string | ReactNode;
}

const Text: FC<TextProps> = ({ as: TextTag = 'p', children }) => (
    <TextTag className={styles['text']}>{children}</TextTag>
);

export default Text;
