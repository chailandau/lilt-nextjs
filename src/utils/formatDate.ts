/**
 * Formats a given timestamp into a string representation of the date.
 *
 * @param timestamp - The timestamp to be formatted.
 * @returns The formatted date string.
 */
export const formatDate = (timestamp: string | number | Date) => {
    const date = new Date(timestamp);

    return date.toLocaleDateString('en-US', {
        year: 'numeric',
        month: 'long',
        day: '2-digit'
    });
};
