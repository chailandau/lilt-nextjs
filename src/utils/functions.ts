/**
 * Determine if provided image source URL is an SVG.
 *
 * @param {string} sourceUrl - The URL of the image to check.
 * @return {boolean} True if the URL ends with ".svg", false otherwise.
 */
export const isSvg = (sourceUrl: string): boolean => sourceUrl.endsWith('.svg');

