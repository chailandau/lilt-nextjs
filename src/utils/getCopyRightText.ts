/**
 * Replaces all occurrences of "%CURRENT_YEAR%" in the given string with the current year. If the string is empty, returns a default copyright text.
 *
 * @param str - The string to be modified.
 * @returns The modified string with all occurrences of "%CURRENT_YEAR%" replaced with the current year.
 */

const isLimg = process.env.NEXT_PUBLIC_THEME === 'green-theme';

export const getCopyrightText = (str?: string | null): string => {
    if (!str) {
        str = `Copyright Â© %CURRENT_YEAR% ${
            isLimg ? 'Long Island Mini Golf' : 'Long Island Laser Tag'
        }. All Rights Reserved.`;
    }
    const currentYear = new Date().getFullYear().toString();

    return str.replaceAll('%CURRENT_YEAR%', currentYear);
};
