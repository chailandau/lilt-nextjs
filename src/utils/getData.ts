import { graphQLClient } from '@/api/graphqlClient';
import { Query } from '@/api/graphqlTypes';

/**
 * Retrieves data from the API using the provided GraphQL query.
 *
 * @param QUERY - The GraphQL query to be executed.
 * @param slug - (Optional) The slug to be used as a variable in the query.
 * @returns A promise that resolves to the data retrieved from the API.
 */
export const getData = async (QUERY: string, slug: string | null = null) => {
    try {
        const variables = slug !== null ? { slug } : {};

        return await graphQLClient.request<Query>(QUERY, variables);
    } catch (error) {
        console.error('GraphQL request error:', error);
        throw error;
    }
};
