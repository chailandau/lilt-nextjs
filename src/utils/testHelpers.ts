import { fireEvent, render } from '@testing-library/react';
import { axe } from 'jest-axe';
import { ReactElement, cloneElement } from 'react';

type Component = ReactElement;

interface TestComponent {
    component: Component;
}

interface TestRole extends TestComponent {
    role: string;
}

interface TestTag extends TestComponent {
    htmlTag: string;
}

interface TestPropOptions extends TestTag {
    propName: string;
    propOptions: readonly string[];
}

interface TestTags extends TestComponent {
    tags: readonly string[];
}

interface TestRenderText extends TestRole {
    text: string;
}

/**
 * Maps through a list of prop options from a provided array, testing that each rendered component contains the specified prop option.
 *
 * @param component - Component to test.
 * @param propName - Name of prop to pass options to.
 * @param propOptions - Array of prop options to map through.
 * @param htmlTag - HTML tag of the component.
 */
export const testPropOptions = ({
    component,
    propName,
    propOptions,
    htmlTag
}: TestPropOptions): void => {
    propOptions.forEach((propOption) => {
        it(`has '${propOption}' class`, () => {
            const { container } = render(
                cloneElement(component, { [propName]: propOption })
            );
            expect(container.querySelector(htmlTag)).toHaveClass(propOption);
        });
    });
};

/**
 * Tests that `className` is applied to component if it exists.
 *
 * @param component - Component to test.
 * @param htmlTag - HTML tag of the component.
 */
export const testAppliesClass = ({ component, htmlTag }: TestTag): void => {
    it('applies class if it exists', () => {
        const { container } = render(
            cloneElement(component, { className: 'test' })
        );
        expect(container.querySelector(htmlTag)).toHaveClass('test');
    });
};

/**
 * Maps through a list of tags from a provided array, testing that each rendered component
 * contains the specified tag.
 *
 * @param component - Component to render.
 * @param tags - Array of tags to map through.
 */
export const testTags = ({ component, tags }: TestTags): void => {
    tags.forEach((tag) => {
        it(`has '${tag}' tag`, () => {
            const { container } = render(cloneElement(component, { as: tag }));
            expect(container.querySelector(tag)).toBeDefined();
        });
    });
};

/**
 * Tests that component renders with correct text.
 *
 * @param component - Component to render.
 * @param text - Text to render.
 */
export const testRenderText = ({ component, text }: TestRenderText): void => {
    it('renders with correct text', () => {
        const { container } = render(component);
        expect(container.textContent).toBe(text);
    });
};

/**
 * Tests that children render correctly inside a Higher Order Component (HOC) component.
 *
 * @param component - Component to test.
 */
export const testRenderChildren = ({ component }: TestComponent): void => {
    it('renders children inside HOC component', () => {
        const { getByText } = render(component);
        expect(getByText('Test')).toBeDefined();
    });
};

/**
 * Tests that keydown event is fired when key is pressed.
 *
 * @param component - Component to test.
 * @param role - Component's ARIA role.
 */
export const testKeyDown = ({ component, role }: TestRole): void => {
    it('should call onKeyDown when a key is pressed', () => {
        const onKeyDownMock = jest.fn();

        const { getByRole } = render(
            cloneElement(component, {
                onKeyDown: onKeyDownMock
            })
        );
        const element = getByRole(role);
        fireEvent.keyDown(element, { key: 'Enter', code: 13 });
        expect(onKeyDownMock).toHaveBeenCalledTimes(1);

        const eventObject = onKeyDownMock.mock.calls[0][0];
        expect(eventObject).toBeDefined();
        expect(eventObject.key).toBe('Enter');
    });
};

/**
 * Tests that component matches snapshot and does not have axe violations.
 *
 * @param component - Component to test.
 */
export const testAxeAndSnapshot = ({ component }: TestComponent) => {
    it('matches snapshot', () => {
        const { asFragment } = render(component);
        expect(asFragment()).toMatchSnapshot();
    });
    it('does not have axe violations', async () => {
        const { container } = render(component);
        expect(await axe(container)).toHaveNoViolations();
    });
};
