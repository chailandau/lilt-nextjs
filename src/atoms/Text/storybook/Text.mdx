import { Canvas, Meta, ArgTypes } from '@storybook/blocks';

import * as TextStories from './Text.stories';

<Meta of={TextStories} />

# Text Atom

<Canvas of={TextStories.Default} />

<br />

## children

Text is supplied as a child to the `Text` atom.

## as

You can choose to use the text atom as a `p` or `span` tag for flexibility. If the `as` prop is not defined, the tag will default to `p`.

<br />
<br />
<br />

## Props Table

<ArgTypes />
