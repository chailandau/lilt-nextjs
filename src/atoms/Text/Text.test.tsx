import { composeStory } from '@storybook/react';

import Meta, { Default } from './storybook/Text.stories';
import { textTags } from './Text';

import { testAxeAndSnapshot, testTags } from '@/utils/testHelpers';

const Text = composeStory(Default, Meta);

const textComponent = {
    component: <Text />
};

describe('Text', () => {
    testTags({ ...textComponent, tags: textTags });
    testAxeAndSnapshot(textComponent);
});
