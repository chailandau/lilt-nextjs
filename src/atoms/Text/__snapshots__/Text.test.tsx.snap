// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text matches 'p' snapshot 1`] = `
<p
  className="text"
>
  This is text.
</p>
`;

exports[`Text matches 'span' snapshot 1`] = `
<span
  className="text"
>
  This is text.
</span>
`;

exports[`Text matches snapshot 1`] = `
<p
  className="text"
>
  This is text.
</p>
`;
